packages/loader/container-loader/src/deltaManager.ts

```diff
@@ -783,12 +812,36 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
                     }
                 }
             } else if (message.sequenceNumber !== this.lastQueuedSequenceNumber + 1) {
+                // TODO: might not need to do anything here (just circles back to else below)
                 this.pending.push(message);
                 this.fetchMissingDeltas(reason, message.sequenceNumber);
             } else {
+                // Client ops: MSN has to be lower than sequence #, as client can continue to send ops with same
+                // reference sequence number as this op.
+                // System ops (when no clients are connected) are the only ops where equation is possible.
+                const diff = message.sequenceNumber - message.minimumSequenceNumber;
+                if (diff < 0 || diff === 0 && message.clientId !== null) {
+                    // TODO: this hit in e2e tests
+                    throw new DataCorruptionError("MSN has to be lower than sequence #",
+                        extractSafePropertiesFromMessage(message));
+                }
+
                 this.lastQueuedSequenceNumber = message.sequenceNumber;
                 this.previouslyProcessedMessage = message;
-                this._inbound.push(message);
+                if (message.type === GroupedBatchOpType) {
+                    // TODO: do we want to process the grouped batch op?
+                    for (const subMessage of message.contents as IDocumentMessage[]) {
+                        this.updateLatestKnownOpSeqNumber(++this.virtualSequenceNumber);
+                        this._inbound.push({
+                            ...message, // This will override the property difference between ISequencedDocumentMessage and IDocumentMessage
+                            ...subMessage,
+                            sequenceNumber: this.virtualSequenceNumber,
+                        });
+                    }
+                } else {
+                    this.updateLatestKnownOpSeqNumber(++this.virtualSequenceNumber);
```

Comment:
It will be very hard to follow code and make it right if you start mixing virtualized and non-virtualized values here.
At a minimum, I'd suggest starting using branded types, such that it's impossible to accidently mix and match them and provide conversion functions between types (you will need conversion back and forth - you need herr to map referenceSequenceNumber, clientSequenceNumber as well as sequenceNumber to virtualized numbers).

But I'd suggest using Delta Manager adapter approach. Leveraging branded types on top of it would be a plus, but maybe it's secondary step

Reply:
Even if you make it work by injecting code here and there across DM, it will be very hard to follow this code / understand - the risk of future developers making mistakes would increase substantially. An adapter allows isolation / testability, and results in most developers to not need to touch that code ever.